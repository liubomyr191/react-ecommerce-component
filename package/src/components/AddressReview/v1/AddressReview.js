import React, { Component } from "react";
import PropTypes from "prop-types";
import uniqueId from "lodash.uniqueid";
import { Form } from "reacto-form";
import styled from "styled-components";
import { withComponents } from "@reactioncommerce/components-context";
import { applyTheme, addTypographyStyles, CustomPropTypes } from "../../../utils";

// component styles
const WarningMessage = styled.div`
  ${addTypographyStyles("WarningMessage", "labelText")};
  border-color: ${applyTheme("AddressReview.warningMessageBorderColor")};
  border-style: ${applyTheme("AddressReview.warningMessageBorderStyle")};
  border-width: ${applyTheme("AddressReview.warningMessageBorderWidth")};
  background-color: ${applyTheme("AddressReview.warningMessageBackgroundColor")};
  color: ${applyTheme("AddressReview.warningMessageColor")};
  padding-bottom: ${applyTheme("AddressReview.warningMessagePaddingBottom")};
  padding-left: ${applyTheme("AddressReview.warningMessagePaddingLeft")};
  padding-right: ${applyTheme("AddressReview.warningMessagePaddingRight")};
  padding-top: ${applyTheme("AddressReview.warningMessagePaddingTop")};
  white-space: pre-wrap;
`;

const ENTERED = "entered";
const SUGGESTED = "suggested";

class AddressReview extends Component {
  static propTypes = {
    /**
     * Address entered
     */
    addressEntered: CustomPropTypes.address.isRequired,
    /**
     * Address validations address suggestion
     */
    addressSuggestion: CustomPropTypes.address.isRequired,
    /**
     * You can provide a `className` prop that will be applied to the outermost DOM element
     * rendered by this component. We do not recommend using this for styling purposes, but
     * it can be useful as a selector in some situations.
     */
    className: PropTypes.string,
    /**
     * If you've set up a components context using @reactioncommerce/components-context
     * (recommended), then this prop will come from there automatically. If you have not
     * set up a components context or you want to override one of the components in a
     * single spot, you can pass in the components prop directly.
     */
    components: PropTypes.shape({
      /**
       * Pass either the Reaction Address component or your own component that
       * accepts compatible props.
       */
      Address: CustomPropTypes.component.isRequired,
      /**
       * Pass either the Reaction Checkbox component or your own component that
       * accepts compatible props.
       */
      Checkbox: CustomPropTypes.component.isRequired,
      /**
       * Pass either the Reaction Field component or your own component that
       * accepts compatible props.
       */
      Field: CustomPropTypes.component.isRequired
    }).isRequired,
    /**
     * Is data being saved
     */
    isSaving: PropTypes.bool,
    /**
     * Form name
     */
    name: PropTypes.string,
    /**
     * Form submit event callback
     */
    onSubmit: PropTypes.func,
    /**
     * The selected address option
     */
    value: PropTypes.string,
    /**
     * Warning message to display above the form
     */
    warningMessage: PropTypes.string
  };

  static defaultProps = {
    isSaving: false,
    value: SUGGESTED,
    // eslint-disable-next-line
    warningMessage: `The address you entered may be incorrect or incomplete.\n\nPlease review our suggestion below, and choose which version youâ€™d like to use. Errors are shown in red.`
  };

  _form = null;

  uniqueInstanceIdentifier = uniqueId("AddressReviewForm_");

  submit = () => this._form.submit();

  handleSubmit = (value) => {
    const { addressEntered, addressSuggestion, onSubmit } = this.props;
    const selectedAddress = value === ENTERED ? addressEntered : addressSuggestion;
    onSubmit(selectedAddress);
  };

  render() {
    const {
      addressEntered,
      addressSuggestion,
      className,
      components: { Address, SelectableList },
      isSaving,
      onSubmit,
      value,
      warningMessage
    } = this.props;

    const options = [
      {
        id: `${ENTERED}_${this.uniqueInstanceIdentifier}`,
        detail: <Address address={addressEntered} />,
        label: "Entered Address:",
        value: ENTERED
      },
      {
        id: `${SUGGESTED}_${this.uniqueInstanceIdentifier}`,
        detail: <Address address={addressSuggestion} />,
        label: "Suggested Address:",
        value: SUGGESTED
      }
    ];

    return (
      <div className={className}>
        <WarningMessage>{warningMessage}</WarningMessage>
        <Form
          ref={(formEl) => {
            this._form = formEl;
          }}
          onSubmit={onSubmit}
        >
          <SelectableList
            isHorizontal
            isStacked
            options={options}
            name="AddressReview"
            value={value}
            isReadOnly={isSaving}
          />
        </Form>
      </div>
    );
  }
}

export default withComponents(AddressReview);
