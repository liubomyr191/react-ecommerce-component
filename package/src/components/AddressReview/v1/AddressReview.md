### Overview
#### Usage
The `AddressReview` is a simple form that requires two address objects, `addressEntered` and `addressSuggestion` will be used to create the `SelectableList` options. The "selected" address will be returned on submit.

```jsx
const addressEntered = {
  address1: "7742 Hwy 25",
  address2: "",
  country: "US",
  city: "Belle Chasse",
  fullName: "Salvos Seafood",
  postal: "70047",
  region: "LA",
  phone: "(504) 393-7303"
};

const addressSuggestion = {
  address1: "7742 Hwy 23",
  address2: "",
  country: "US",
  city: "Belle Chasse",
  fullName: "Salvos Seafood",
  postal: "70037",
  region: "LA",
  phone: "(504) 393-7303"
};

const props = { addressEntered, addressSuggestion };

<AddressReview {...props} />
```

##### Value
By default the suggested address is selected but it can easily be changed by pasing a `value` string of "entered" to pre select the entered address.
```jsx
const addressEntered = {
  address1: "7742 Hwy 25",
  address2: "",
  country: "US",
  city: "Belle Chasse",
  fullName: "Salvos Seafood",
  postal: "70047",
  region: "LA",
  phone: "(504) 393-7303"
};

const addressSuggestion = {
  address1: "7742 Hwy 23",
  address2: "",
  country: "US",
  city: "Belle Chasse",
  fullName: "Salvos Seafood",
  postal: "70037",
  region: "LA",
  phone: "(504) 393-7303"
};

const props = { addressEntered, addressSuggestion, value: "entered" };

<AddressReview {...props} />
```

##### Warning Message
The displayed warning message can be changed by providing a `warningMessage` string as a prop. The text of this element respects line breaks via `\n` within the strings text.
```jsx
const addressEntered = {
  address1: "46 Avenue M",
  address2: "",
  country: "HT",
  city: "Port-au-Prince",
  fullName: "Yanvalou",
  postal: "6111",
  region: "OU",
  phone: "509 38 01 7051"
};

const addressSuggestion = {
  address1: "46 Avenue N",
  address2: "",
  country: "HT",
  city: "Port-au-Prince",
  fullName: "Yanvalou",
  postal: "6110",
  region: "OU",
  phone: "509 38 01 7051"
};

const props = { 
  addressEntered, 
  addressSuggestion, 
  warningMessage: "Adrès ou te antre a ka kòrèk oswa enkonplè.\n\nTanpri revize sijesyon nou anba a, epi chwazi ki vèsyon ou ta renmen itilize. Erè yo montre nan wouj." 
};

<AddressReview {...props} />
```

#### Example implamentation
Simple `AddressReview` example.
```jsx
const addressEntered = {
  address1: "7742 Hwy 25",
  address2: "",
  country: "US",
  city: "Belle Chasse",
  fullName: "Salvos Seafood",
  postal: "70047",
  region: "LA",
  phone: "(504) 393-7303"
};

const addressSuggestion = {
  address1: "7742 Hwy 23",
  address2: "",
  country: "US",
  city: "Belle Chasse",
  fullName: "Salvos Seafood",
  postal: "70037",
  region: "LA",
  phone: "(504) 393-7303"
};

class AddressExample extends React.Component {
  constructor(props) {
    super(props);
    this._addressForm;
    this.bindForm.bind(this);
  }

  bindForm(formEl) {
    if (formEl) {
      this._addressReview = formEl;
    }
  }

  render() {
    return (
      <div>
        <AddressReview
          ref={(formEl) => { this.bindForm(formEl) }}
          onSubmit={(selectAddress) => console.log("Address selected", selectAddress)}
          addressEntered={addressEntered}
          addressSuggestion={addressSuggestion}
        />
        <Button onClick={() => this._addressReview.submit()}>Submit</Button>
      </div>
    );
  }
}

<AddressExample />
```

### Theme

Assume that any theme prop that does not begin with "rui" is within `rui_components`. See [Theming Components](./#!/Theming%20Components).

| Theme Prop           | Default | Description                                                                              |
| -------------------- | ------- | ---------------------------------------------------------------------------------------- |
| `AddressReview.warningMessageBorderColor` | yellow400   | Warning message border color |
| `AddressReview.warningMessageBorderStyle` | solid   | Warning message border style |
| `AddressReview.warningMessageBorderWidth` | 1px   | Warning message border width |
| `AddressReview.warningMessageBackgroundColor` | yellow100   | Warning message background color |
| `AddressReview.warningMessageColor` | yellow600  | Warning message color |
| `AddressReview.warningMessagePaddingBottom` | 14px   | Warning message bottom padding |
| `AddressReview.warningMessagePaddingLeft` | 20px   | Warning message left padding |
| `AddressReview.warningMessagePaddingRight` | 20px   | Warning message right padding |
| `AddressReview.warningMessagePaddingTop` | 14px   | Warning message top padding |
| `AddressReview.formSpacingTop` | 40px   | Space between form and warning message |

#### Typography

- The warning message text uses `labelText` style with `rui_components.WarningMessage` override
