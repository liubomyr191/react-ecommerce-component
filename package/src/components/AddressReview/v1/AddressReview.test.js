import React from "react";
import renderer from "react-test-renderer";
// import { shallow } from "enzyme";
import AddressReview from "./AddressReview";

test("basic snapshot", () => {
  const component = renderer.create(<AddressReview />);

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
