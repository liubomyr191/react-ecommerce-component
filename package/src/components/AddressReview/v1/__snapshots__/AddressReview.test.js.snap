// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic snapshot with all props 1`] = `
.c0 {
  margin-top: 40px;
}

<div
  className={undefined}
>
  InlineAlert(undefined)
  <div
    className="c0"
  >
    <div
      className={null}
      style={Object {}}
    >
      SelectableList(undefined)
    </div>
  </div>
</div>
`;

exports[`basic snapshot with only required props 1`] = `
.c0 {
  margin-top: 40px;
}

<div
  className={undefined}
>
  InlineAlert(undefined)
  <div
    className="c0"
  >
    <div
      className={null}
      style={Object {}}
    >
      SelectableList(undefined)
    </div>
  </div>
</div>
`;
